### PR TITLE
[Sampling Replay 1/8] Add sampling-support log-prob primitives

### PR DESCRIPTION
**Pull request by nanjiangwill**
## Summary

- Add an optional per-token sampling-support mask primitive for actor log-probability computation.
- Preserve the public get_responses() two-tuple contract while deriving context-parallel response-row indices only when support replay needs them.
- Apply an optional support mask through the existing true-on-policy and fused tensor-parallel cross-entropy paths.

No production caller supplies a sampling mask in this PR, so rollout and training behavior remain unchanged.

## Why

The trainer needs a correct way to normalize actor log probabilities over the same sampling support used during rollout. This primitive intentionally landed before capture was activated: rollout log probabilities must not switch to a restricted distribution until actor scoring can use the same support.

The logical input follows rollout data directly: one list of supported token IDs per response token. The enabled path selects response rows owned by the current context-parallel rank, flattens those supports once, and builds one tensor-parallel-local boolean mask. The disabled path does not build response indices or allocate a mask.

## Sampling-support replay series

| Part | PR | Responsibility | State |
|---:|---|---|---|
| 1/8 | #2200 | TP/CP-aware masked log-probability primitive | Merged |
| 2/8 | #2595 | Compact support representation and transport | Merged |
| 3/8 | #3327 | Actor-only Megatron and FSDP integration | Open |
| 4/8 | #3328 | Native SGLang support capture | Open |
| 5/8 | #3329 | Session and agentic preservation | Open |
| 6/8 | #3330 | User-facing validation | Open |
| 7/8 | #3331 | GPU E2E | Open |
| 8/8 | #3332 | Documentation | Open |

External prerequisite: #3321 provides the SGLang v0.5.20 base. It is owned by another contributor and is not part of this numbered series.

#2596 is the [DO NOT MERGE] review umbrella for the complete stack.

## Validation

- Repository lint and formatting checks passed on every changed file.
- Focused CPU checks cover nested-support selection, tensor-parallel vocabulary filtering, validation, and empty local response shards.

## Acknowledgements

Thank you to [Zilin Zhu](https://github.com/zhuzilin) for adding top-p masking support to slime in [THUDM/slime#2102](https://github.com/THUDM/slime/pull/2102).

### CODE REVIEW

File: miles/backends/training_utils/loss_hub/logit_processors.py [hunk 1]
```diff
@@ -139,6 +191,8 @@ def get_log_probs_and_entropy(
     entropy_requires_grad: bool = True,
     non_loss_data: bool = True,
     max_seq_lens: list[int] | None = None,
+    rollout_sampling_mask_ids: Sequence[Sequence[int] | torch.Tensor] | None = None,
```
Comment (guapisolo) [inline miles/backends/training_utils/loss_hub/logit_processors.py]: Why didn't directly use `list[list[int]]` as data format? I think it's not very easy to understand.

Comment (guapisolo) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: Codex said mooncake can support this by some tiny modifications

Comment (nanjiangwill) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: addressed

Comment (guapisolo) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: sry what i mean is, why you use csr (flatten tensor + offset) instead of direct storage, (like single list list list int or list list tensor)

Comment (guapisolo) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: i think the original shape can reduce some reading overhead lol

Comment (nanjiangwill) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: for logical model the nested structure is better

but for data transport the current csr is better

will update a bit

Comment (guapisolo) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: I have a small PR, can you check whether that looks good to you

Comment (guapisolo) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: done. here https://github.com/nanjiangwill/miles/pull/1

Comment (nanjiangwill) [reply miles/backends/training_utils/loss_hub/logit_processors.py]: merged

**Review by guapisolo — APPROVED** [review]

**Review by guapisolo — APPROVED** [review]
LGTM now. wait code owner's review.

**Review by yueming-yuan — APPROVED** [review]

**Comment by yueming-yuan** [comment]
please check the CI failure

**Comment by guapisolo** [comment]
> please check the CI failure

seems like introduced by rdt PR instead of this one.. will submit a follow-up PR to fix.
